### ISSUE TITLE
Ably and ClientWebsocket

### ISSUE THREAD

**Issue by riccardobecker** [ably/ably-dotnet#1257]
Hi Ably fans,

I was looking into PaaS solution that can handle thousands/millions of websocket connections. 

I stumbled upon Ably but I cannot find any hints on how to use the ClientWebsocket (or some other websocket client SDK) directly with the Ably endpoints. We are servicing devices that are out in the field and they can only connect using ws or wss protocol. The only thing they will do is add some segments to the base URL to identify themselves with our backend. Ideally, we would redirect those clients to some "websocket" endpoint of our Ably instance so we have a stable websocket layer that has an excellent SLA. Also we have some obvious downtime right now when we update our ws layer.

Is there a possiblity to achieve this scenario?

regards,
Riccardo

**Comment by sacOO7** [comment]
@riccardobecker We have official client SDK's that follow ably custom protocol specifications built on top of WebSocket. So, if you want to use `ably`, you have to integrate those client SDK's into your devices ( you will need to update the code accordingly). I can guarantee, that ably is a fantastic PAAS solution, but there is a one-time effort to integrate client SDK into your client application. 

You can also use other client SDK's, but they will use other protocols. You can check official and unofficial client SDK's here ->
https://ably.com/download

**Comment by riccardobecker** [comment]
Thanks for the reply. The problem is that I don't control the clients and they follow a certain protocol which makes the "clean" websocket connection the only way to go. There is no way I can update the client's firmware whatsoever.

**Comment by sacOO7** [comment]
@riccardobecker makes sense! I have one last question, How big is your current server code? If you had a choice, would you migrate your server logic to a service that can scale to millions of connections? Even rewrite in a different programming language if needed (mostly JS or Python)? What's your current programming language at the backend?

**Comment by riccardobecker** [comment]
the code is not the issue, it's merely the fact that we need a layer that will not update frequently (or not at all) that handles the websocket connections and then offloads to whatever async layer like mqtt.

**Comment by sacOO7** [comment]
@riccardobecker seems having a custom lightweight websocket server code as a part of that layer will help make it work. I don't think ably supports it yet : (

**Comment by riccardobecker** [comment]
No Ably doesn't support this scenario too bad!

**Comment by mattheworiordan** [comment]
@riccardobecker the closest thing we have is native SSE support, see https://ably.com/docs/protocols/sse.  Our websocket protocol is feature rich, and we have no roadmap plans to make raw websockets available I am afraid given the reason we exist is because our SDKs provide a rich API feature set to developers so that they don't need to operate at a websocket/protocol level. Someone in our solution engineering team may be able to offer some advice though to see what we could do to help. You can set up a call with one of them by reaching out at https://ably.com/contact.  I hope we can help find a solution.

**Comment by sacOO7** [comment]
@riccardobecker let us know if you have more updates on this, otherwise we will be closing the issue
